cap-primitives/src/std/fs/open_manually.rs

```diff
@@ -60,11 +60,28 @@ impl<'path_buf> CanonicalPath<'path_buf> {
         }
     }
 
-    fn pop(&mut self) {
+    fn pop(&mut self) -> bool {
         #[cfg(debug_assertions)]
         self.debug.pop();
         if let Some(path) = &mut self.path {
-            path.pop();
+            path.pop()
+        } else {
+            true
+        }
+    }
+
+    /// The complete canonical path has been scanned. Set `path` to `None`
+    /// so that it isn't cleared when `self` is dropped.
+    fn complete(&mut self) {
+        self.path = None;
+    }
+}
+
+impl<'path_buf> Drop for CanonicalPath<'path_buf> {
+    fn drop(&mut self) {
+        if let Some(path) = &mut self.path {
+            path.clear();
+            self.path = None;
         }
     }
 }
```

Comment:
I feel a bit uneasy about this. `self.path` is mutable reference so I'd imagine the owner should handle clearing, deallocating, etc. Besides, is this on-drop behaviour actually needed at all anywhere?

Reply:
`canonicalize` can succeed even if the last component of the path is not accessible. `open_manually` will `Err` in that case, so `canonicalize` needs a separate way to determine if it got a complete path or not. Clearing the `canonical_path` is how the current patch indicates that the complete path is not available.

It comes from trying to keep `canonicalize`'s impact on `open_manually` minimal, because none of its other users care about the canonical path. With the current path, all they have to do is pass `None`. And then `canonicalize` passes a `&mut PathBuf` which then comes back with either the desired result or empty.

I've now added some comments to clarify this. Does this clarify what's going on?

Reply:
Yep, thanks!

In a future effort, perhaps it'd be worth refactoring it into a state machine or some such? This way, we could signal failure by encoding a state, and the user would inquire about the state to work out if the processing was successful or not? This way, we wouldn't have to touch the ref to `PathBuf` in `drop`.